### ISSUE TITLE
Windows support for distributed training

### ISSUE THREAD

**Issue by Ownmarc** [pytorch/pytorch#31363]
## 🚀 Feature
Windows support for distributed training (multiple GPUs on the same host)

## Motivation

I use distributed training with Pytorch on Linux and it is really easy and works well. I would like the same for Windows. I work alot with images and I find it easier to debug and see what I am doing in general on Windows and it would be great if I didn't have to dualboot to Linux when I want to multi-gpu train my models.

## Pitch

Being able to use distributed training on Windows. I know it wasn't requested much by users, but lets have this feature request open and see if others would like it.

## Alternatives

Linux

## Additional context

-

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @peterjc123

**Comment by wallcuber** [comment]
Do need windows support for distributed training. I can't train [ultralytics/yolov3](https://github.com/ultralytics/yolov3/blob/master/train.py) with my 4 GPUs. It limits the scale of my dataset and the size of input image.

**Comment by pietern** [comment]
For whoever is interested in working on this:

1. Add Windows support to [Gloo](https://github.com/facebookincubator/gloo). On the I/O side, Gloo already has a [libuv](https://libuv.org/) transport, which should work fine on Windows. It's just the CMake build system that needs to be adapted to work on Windows.
2. Once that is done, the PyTorch build system should be updated to compile with `USE_DISTRIBUTED` when compiling on Windows. I'm sure there are other things besides the build system that come up, but I can't tell what they would be...

Glad to help if anyone is interested in working on this, but this is not a priority for the team.

**Comment by y78h11b09** [comment]
> For whoever is interested in working on this:
> 
> 1. Add Windows support to [Gloo](https://github.com/facebookincubator/gloo). On the I/O side, Gloo already has a [libuv](https://libuv.org/) transport, which should work fine on Windows. It's just the CMake build system that needs to be adapted to work on Windows.
> 2. Once that is done, the PyTorch build system should be updated to compile with `USE_DISTRIBUTED` when compiling on Windows. I'm sure there are other things besides the build system that come up, but I can't tell what they would be...
> 
> Glad to help if anyone is interested in working on this, but this is not a priority for the team.

I  need windows support for distributed training, can you help me to implement the distiributed training on windows ?

**Comment by mrshenli** [comment]
Hey @y78h11b09, @wallcuber, @Ownmarc,

Windows supports for DistributedDataParallel (only with FileStore , i.e., init_method="file://...") are landed in both master and [1.7 release branch](https://github.com/pytorch/pytorch/tree/release/1.7). Please feel free to try it and let us know your thoughts. Thanks!

I am closing this issue. Let's discuss and the progress of remaining features in #42095.
